### PR TITLE
HDMI hotplugging

### PR DESCRIPTION
**Pull request by knuton**
WIP

## Checklist

-   [ ] Changelog updated
-   [ ] Code documented
-   [ ] User manual updated

### CODE REVIEW

**Comment by knuton** [comment]
The current setup seems to cause the boot to hang, possibly the udev rule triggers while X is not yet up and running.

Alternative idea: Run the script as a one-shot systemd service, depend on `multi-user.target` (or something that guarantees X server is up), and restart the service from the udev rule when the event triggers.

**Comment by knuton** [comment]
Closing this for now. We realized we also need to deal with the hotplugging in the kiosk, which currently has no notion of the resolution changing and needing to resize.

**Comment by yfyf** [comment]
Reopening this, because it is a good starting point, just need to address a few extra points.

File: application.nix [hunk 1]
```diff
@@ -126,6 +123,28 @@ rec {
...
+      # Monitor hotplugging
+      services.udev.extraRules = ''
+        ACTION=="change", SUBSYSTEM=="drm", RUN+="${pkgs.systemd}/bin/systemctl restart select-display.service"
+      '';
+      systemd.services."select-display" = {
+        description = "Select best display to output to";
+        serviceConfig.Type = "oneshot";
+        serviceConfig.ExecStart = "${pkgs.bash}/bin/bash ${./application/select-display.sh}";
+        serviceConfig.User = "play";
+        serviceConfig.After = "multi-user.target";
```
Comment (yfyf) [inline application.nix:135]: I would expect this to be linked (through some rule) to the `graphical.target` instead? Also, I don't think the `After=`  will actually wait for the `multi-user.target`, so the `udev` rule will launch this too  early?

File: application.nix [hunk 2]
```diff
@@ -126,6 +123,28 @@ rec {
...
+        environment = {
+          XAUTHORITY = "${config.users.users.play.home}/.Xauthority";
+          DISPLAY = ":0";
+        };
+        path = with pkgs; [
+          gnugrep
+          gawk
+          xorg.xrandr
+        ];
+        wantedBy = [ "multi-user.target" ];
```
Comment (yfyf) [inline application.nix:145]: Ditto, probably best to remove it as @guyonvarch did in 85f5efa.

**Comment by yfyf** [comment]
> We realized we also need to deal with the hotplugging in the kiosk, which currently has no notion of the resolution changing and needing to resize.

My proposal would be _not_ to do anything in the kiosk and simply restart it (or the whole xsession) after the screen is re-plugged? This is a one-off event that does not have to be handled very gracefully, but it would save us a lot of debugging when Qt fails to rescale. The only issue if the user hotplugs a screen mid-setup or mid-game with unsaved input, but that is kinda... unlikely?

File: application.nix (hunk 1 above)
Comment (knuton) [reply application.nix:135]: Honestly I'd need a moment to refresh my memory on the targets. I dimly remember trying to rely on a target that I then realized would not be present on our minimal system. But I just checked for `graphical.target` inside a VM PlayOS and it's there, so plausibly yes.

Regarding the ordering conditions, I understand `After=` to actually wait:

> After= ensures [...] that the listed unit is fully started up before the configured unit is started.

https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#Before=

**Comment by knuton** [comment]
> My proposal would be _not_ to do anything in the kiosk and simply restart it (or the whole xsession) after the screen is re-plugged? This is a one-off event that does not have to be handled very gracefully, but it would save us a lot of debugging when Qt fails to rescale. The only issue is if the user hotplugs a screen mid-setup or mid-game with unsaved input, but that is kinda... unlikely?
> 
> Edit: alternatively, Qt provides the [availableGeometryChanged](https://doc.qt.io/qt-6/qscreen.html#availableGeometry-prop) signal, but if we want to respond to it _without_ restarting, then we have propagate this to all child widgets, which might be non-trivial / hard to test?

My proposal would be to try with exactly the kind of mechanism you cite, which may or may not work well considering we have no proper window manager running. When we understand if/why it's hard to do, we can consider restarting the kiosk instead.

Restarting the kiosk/session may be problematic, if there is a problem with a cable/adapter/plug that causes it to reconnect at intervals.

File: application.nix (hunk 1 above)
Comment (yfyf) [reply application.nix:135]: > Regarding the ordering conditions, I understand After= to actually wait <..>

Yes, but only if both are being started at the same time, i.e. it is a "soft" dependency, from the same man page:

> If unit foo.service contains the setting Before=bar.service and **both units are being started**, bar.service's start-up is delayed until foo.service has finished starting up. After= is the inverse of Before=,

Comment (knuton) [reply application.nix:135]: I see how you read it, though I wouldn't have read it like that.

What if we use both `After=` and `Requires=`?

**Comment by yfyf** [comment]
Moving further development to #199

**Pull request closed without merging**